### PR TITLE
feat(schema): Make section IDs optional and fix movement detection with marker changes

### LINKED ISSUE #51: feat(schema): Make section IDs optional in input YAML
## Description

Make section `id` fields optional in the OpenSpec schema while maintaining auto-generation in Pydantic models. This simplifies the input format since IDs are not used for matching sections across versions (markers are used instead).

## Context

Currently, section IDs are:
- **Required** in the OpenSpec schema (`legal_document_spec.yaml`)
- **Auto-generated** by Pydantic if not provided (`default_factory=lambda: str(uuid4())`)
- **Not used for matching** - markers are the primary identifiers for diffing
- **Only used for tracking/reference** - stored in `DiffResult.section_id` and `id_path` fields but never compared across versions

This creates a contradiction: the schema requires IDs, but Pydantic can generate them. More importantly, IDs serve no functional purpose in the diffing logic - they're just metadata that gets passed through.

## Unknowns / Clarifications Needed

- [ ] Should we keep `section_id` and `id_path` fields in `DiffResult` for backward compatibility, or remove them entirely?
- [ ] Are there any external systems or integrations that depend on stable section IDs?
- [ ] Should we add a migration guide for users who currently provide IDs?

## Success Criteria (Definition of Done)

- [ ] Remove `id` from `required` array in OpenSpec schema (`legal_document_spec.yaml`)
- [ ] Update schema description to indicate IDs are optional and auto-generated
- [ ] Verify Pydantic auto-generation still works correctly
- [ ] All existing tests pass (no functional changes expected)
- [ ] Update documentation to reflect that IDs are optional
- [ ] Update examples to show both with and without explicit IDs
- [ ] Verify UI fallback ID generation still works (uses `section_id` in composite key)

### Unit Tests
- [ ] Test loading document without section IDs (should auto-generate)
- [ ] Test loading document with explicit section IDs (should use provided IDs)
- [ ] Test mixed scenario (some sections with IDs, some without)
- [ ] Verify diffing works correctly with auto-generated IDs
- [ ] Verify `DiffResult.section_id` and `id_path` fields are populated correctly

### Integration Tests
- [ ] Test full validation pipeline (OpenSpec + Pydantic) with missing IDs
- [ ] Test API endpoint accepts documents without section IDs
- [ ] Test CLI accepts documents without section IDs
- [ ] Test formatters output includes section_id (even if auto-generated)

### Regression Tests
- [ ] Verify existing documents with explicit IDs still work
- [ ] Verify UI displays correctly with auto-generated IDs
- [ ] Verify discussions feature works (uses `change.id`, not `section_id`)

## Documentation Updates

- [ ] Update `docs/user/schema_reference.md` - mark `id` as optional
- [ ] Update `AGENTS.md` - clarify that IDs are optional and auto-generated
- [ ] Update `examples/` - add example without explicit IDs
- [ ] Update `README.md` if it mentions required IDs
- [ ] Update docstrings in `models/section.py` to clarify optional nature

### Files to Modify

1. **Schema**: `src/yaml_diffs/schema/legal_document_spec.yaml`
   - Remove `id` from `required` array (line 147)
   - Update description to indicate optional/auto-generated

2. **Documentation**: 
   - `docs/user/schema_reference.md` (line 176)
   - `AGENTS.md` (line 383)
   - Example files

3. **Tests**: 
   - `tests/test_edge_cases.py` - update `test_missing_section_id` to expect success
   - `tests/test_schema_validation.py` - update `test_reject_missing_section_id` to expect success
   - Add new tests for optional ID scenarios

### Technical Details

- Pydantic model already supports auto-generation (`default_factory=lambda: str(uuid4())`)
- No changes needed to diffing logic (uses markers, not IDs)
- No changes needed to UI (uses `change.id`, not `section_id` for discussions)
- Formatters will continue to output `section_id` (will contain auto-generated UUIDs)

### Backward Compatibility

- Documents with explicit IDs will continue to work unchanged
- Documents without IDs will now be accepted (previously rejected by OpenSpec validation)
- `DiffResult.section_id` will contain either provided ID or auto-generated UUID

## Related Issues

- Related: #33 (Document Parsing Layer) - may need to handle ID generation in parser
- Related: #50 (Comment interfaces) - discussions use `change.id`, not `section_id`, so unaffected

## Estimated Effort

**Medium (M)** - 1-2 days
- Schema change: 30 min
- Test updates: 2-3 hours
- Documentation: 1-2 hours
- Testing/verification: 2-3 hours

### PR DESCRIPTION
**Pull request by noamoss**
## Summary

This PR makes section IDs optional in the OpenSpec schema while maintaining auto-generation in Pydantic models, and fixes a bug in section movement detection where sections with different markers but same content were incorrectly detected as removed+added instead of moved.

## Related Issues

Fixes #51

### 1. Make Section IDs Optional (Issue #51)
- Remove `id` from `required` array in OpenSpec schema (`legal_document_spec.yaml`)
- Update schema descriptions to indicate IDs are optional and auto-generated
- Update tests to expect success for missing IDs instead of validation errors
- Add comprehensive tests for optional ID scenarios:
  - Loading documents without IDs (auto-generation)
  - Loading documents with explicit IDs (use provided)
  - Mixed scenarios (some with IDs, some without)
  - Diffing with auto-generated IDs
  - DiffResult.section_id and id_path population
- Update documentation:
  - `schema_reference.md`: Mark `id` as optional
  - `AGENTS.md`: Clarify that IDs are optional and auto-generated
  - `section.py`: Update docstring to clarify optional nature
- Add example document without explicit IDs (`minimal_document_no_ids.yaml`)

### 2. Fix Section Movement Detection with Marker Changes
- Refactor `_find_moved_sections` to match by content similarity (≥0.95) instead of marker-first
- Remove marker requirement for movement detection - now matches by content similarity regardless of marker or title
- Fix DiffResult for SECTION_MOVED to include title and content fields
- Update existing test to expect correct MOVED behavior
- Add new test cases:
  - Root-level marker change (user's specific case: marker "1" → "2")
  - Movement with title change (handled separately)
  - One-to-one matching with multiple candidates
- Update documentation to clarify matching is by content similarity, not marker or title

## Testing

- [x] All 455 tests pass
- [x] Pre-commit hooks pass
- [x] Code coverage: 85%
- [x] No regressions in existing functionality
- [x] Integration tests pass (API, CLI, formatters)
- [x] User's specific case verified (marker "1" → "2" at root level)

## Backward Compatibility

- Documents with explicit IDs continue to work unchanged
- Documents without IDs are now accepted (previously rejected by OpenSpec validation)
- `DiffResult.section_id` contains either provided ID or auto-generated UUID
- Movement detection now correctly handles marker changes (previously a bug)

## Checklist

- [x] Tests added/updated
- [x] Documentation updated
- [x] Examples updated
- [x] All tests pass
- [x] Pre-commit hooks pass
- [x] No breaking changes

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/yaml_diffs/diff.py [hunk 1]
```diff
@@ -161,74 +164,53 @@ def _find_moved_sections(
...
+            new_section, new_marker_path, _ = unmatched_new[new_key]
+
+            # Skip empty content sections
+            if not new_section.content:
+                continue
+
+            # Check content similarity (≥0.95 threshold)
+            # Note: Title changes are handled separately after movement detection
+            similarity = _calculate_content_similarity(old_section.content, new_section.content)
+            if similarity >= 0.95 and similarity > best_similarity:
```
Comment (noamoss) [inline src/yaml_diffs/diff.py]: **Minor improvement suggestion:**

The condition `similarity >= 0.95 and similarity > best_similarity` could be simplified to `similarity > best_similarity` since `best_similarity` is initialized to `0.0`, and we only update it when `similarity >= 0.95`. However, keeping the explicit threshold check makes the intent clearer and acts as a safety check.

**Consideration:** If multiple new sections have the exact same similarity score (e.g., both exactly 0.95) with an old section, the first one encountered will be selected. This is deterministic based on iteration order, which is fine, but worth noting in a comment if non-deterministic behavior would be problematic.

File: src/yaml_diffs/diff.py [hunk 2]
```diff
@@ -161,74 +164,53 @@ def _find_moved_sections(
...
-
-    # Find matches by marker with content similarity check (one-to-one matching)
-    # BUG FIX: Use popleft() to ensure one-to-one matching and prevent cartesian product
-    # when multiple old sections share the same marker with multiple new sections.
-    # Without this fix, if 2 old sections with marker "1" match 2 new sections with
-    # marker "1", all 4 pairings would be created instead of 2 one-to-one matches.
-    # Using deque.popleft() is O(1) instead of list.pop(0) which is O(n).
+    # Match all unmatched sections by content similarity (one-to-one matching)
+    # Process old sections in order and find best match in new sections
     for old_key in list(unmatched_old.keys()):
```
Comment (noamoss) [inline src/yaml_diffs/diff.py]: **Good practice:** Using `list(unmatched_old.keys())` creates a copy of the keys before iteration, which is essential since we're modifying `unmatched_old` during iteration. This prevents `RuntimeError: dictionary changed size during iteration`. Same applies to `list(unmatched_new.keys())` below. Well done!

**Review by noamoss — COMMENTED** [review]
## Code Review Summary

### Overall Assessment
✅ **Approved with minor suggestions** - This PR successfully implements the requested features with good test coverage and documentation. The changes are well-structured and maintain backward compatibility.

### Issues Found (Minor)

1. **Performance Consideration** (`src/yaml_diffs/diff.py:180`)
   - The refactored algorithm changes from O(n) to O(n*m) complexity
   - This is necessary for the fix but worth documenting
   - Should be acceptable for typical use cases

2. **Documentation Clarity** (`src/yaml_diffs/diff.py:155`)
   - Docstring could be clearer about how title changes interact with movement detection
   - Current behavior is correct, but the explanation could be more explicit

3. **Code Clarity** (`src/yaml_diffs/diff.py:203`)
   - The similarity condition is correct but could benefit from a comment about edge cases
   - When multiple sections have identical similarity scores, selection is deterministic but based on iteration order

### Positive Aspects

✅ **Excellent Test Coverage**
- Comprehensive test cases covering:
  - Optional IDs (with/without/mixed scenarios)
  - Movement detection with marker changes
  - Root-level marker changes (user's specific case)
  - Title changes with movement
  - One-to-one matching with duplicate content
- All edge cases appear well-tested

✅ **Backward Compatibility**
- Documents with explicit IDs continue to work unchanged
- Auto-generation works seamlessly
- No breaking changes to the API

✅ **Code Quality**
- Proper use of `list()` to avoid iteration modification errors
- Clean refactoring from marker-based to content-similarity matching
- Well-documented changes in docstrings and comments

✅ **Documentation Updates**
- Schema documentation updated correctly
- AGENTS.md updated with new behavior
- Example file added demonstrating optional IDs
- API reference updated

✅ **Bug Fix Implementation**
- Correctly fixes the movement detection bug
- Properly handles marker changes while maintaining content-based matching
- DiffResult now includes all necessary fields (old_content, new_content, old_title, new_title)

### Suggestions for Future Improvements

1. Consider adding a performance optimization for the common case where markers haven't changed (though this adds complexity)
2. Add a comment explaining the O(n*m) complexity trade-off
3. Consider documenting the deterministic behavior when similarity scores are identical

### Testing Notes
- All 455 tests pass ✅
- Pre-commit hooks pass ✅
- Code coverage: 85% ✅
- No regressions detected ✅

**Recommendation:** Approve and merge after addressing minor documentation suggestions (optional).

**Comment by noamoss** [comment]
## 📋 Review Feedback Addressed

I've applied the following changes based on code review:

### ✅ Changes Made:
- **Improved docstring clarity** about title changes in movement detection ([comment #2573503861](https://github.com/noamoss/yaml_diffs/pull/61#discussion_r2573503861))
  - Clarified that title differences don't prevent movement detection - they're tracked as a separate change type
- **Added performance comment** explaining O(n*m) complexity trade-off ([comment #2573503868](https://github.com/noamoss/yaml_diffs/pull/61#discussion_r2573503868))
  - Documented the complexity change from O(n) to O(n*m) and why it's necessary
  - Noted that this is acceptable for typical use cases
- **Added edge case comment** about identical similarity scores ([comment #2573503856](https://github.com/noamoss/yaml_diffs/pull/61#discussion_r2573503856))
  - Documented deterministic behavior when multiple sections have identical similarity scores

### ✅ Validation:
- All 455 tests passed
- Code formatting verified
- Changes committed locally

### 📦 Next Step:
Pushing changes to remote branch...

---
Ready for re-review! 🚀

File: src/yaml_diffs/diff.py [hunk 3]
```diff
@@ -161,74 +165,64 @@ def _find_moved_sections(
...
+            new_section, new_marker_path, _ = unmatched_new[new_key]
+
+            # Skip empty content sections
+            if not new_section.content:
+                continue
+
+            # Check content similarity (≥0.95 threshold)
+            # Note: Title changes are handled separately after movement detection
+            similarity = _calculate_content_similarity(old_section.content, new_section.content)
+            if similarity >= 0.95 and similarity > best_similarity:
```
Comment (noamoss) [inline src/yaml_diffs/diff.py]: **Edge case consideration:** The condition `similarity >= 0.95 and similarity > best_similarity` means:
- If multiple new sections have similarity exactly 0.95 with an old section, only the first one encountered will be considered (since `0.95 > 0.95` is False)
- If one section has 0.95 and another has 0.96, the 0.96 will be selected (correct behavior)
- If two sections both have 0.96, the first one encountered will be selected

This is deterministic (based on iteration order) and is acceptable behavior. The comment already documents this, which is good!

**Question for consideration:** Should we use `>=` instead of `>` for `best_similarity` to handle the case where the first match has exactly 0.95? Actually, no - the current logic is correct because we want to find the **best** match, and if the first match is 0.95, we should still consider other matches that might be better. The current implementation is correct.

**Review by noamoss — COMMENTED** [review]
## Second Code Review - Detailed Analysis

### Overall Assessment
✅ **Approved** - The PR is well-implemented with comprehensive test coverage. Found one minor issue (outdated comment) and several observations worth noting.

### Issues Found

1. **Outdated Comment** (`src/yaml_diffs/diff.py:439`) ⚠️
   - **Problem:** Comment says "Find moved sections (same marker, different path)" but the function now handles different markers too
   - **Impact:** Minor - misleading documentation
   - **Fix:** Update comment to: `# Find moved sections (different path, possibly different marker)`

### Detailed Observations

#### Algorithm Correctness ✅
- The content-similarity-based matching algorithm is correctly implemented
- One-to-one matching is properly handled by removing matched items from both sets
- Edge cases (empty content, identical similarity scores) are handled correctly
- The O(n*m) complexity is necessary and acceptable for the use case

#### Code Quality ✅
- Proper use of `list()` to avoid iteration modification errors
- Good documentation in docstrings
- Performance considerations are documented
- Edge cases are commented

#### Test Coverage ✅
- Comprehensive tests for optional IDs (with/without/mixed scenarios)
- Tests for movement detection with marker changes
- Tests for one-to-one matching with duplicate content
- Tests for title changes with movement
- All edge cases appear well-covered

#### Potential Considerations (Not Issues)

1. **Content Similarity Algorithm** (`src/yaml_diffs/diff.py:127`)
   - Uses word-based similarity with `split()` on whitespace
   - Works correctly for Hebrew text and handles whitespace normalization
   - This is a pre-existing design choice, not an issue with this PR

2. **Similarity Threshold Edge Cases** (`src/yaml_diffs/diff.py:211`)
   - The condition `similarity >= 0.95 and similarity > best_similarity` is correct
   - Handles cases where multiple sections have identical similarity scores deterministically
   - Well-documented in code comments

3. **UUID Auto-generation** (`src/yaml_diffs/models/section.py:39`)
   - Correctly uses `default_factory=lambda: str(uuid4())` to ensure unique IDs
   - Proper Pydantic pattern for optional fields with generated defaults

### Positive Highlights

✅ **Excellent Test Suite**
- `test_section_moved_multiple_same_content` validates one-to-one matching
- `test_section_moved_root_level_marker_change` tests the user's specific case
- `test_load_document_mixed_ids` covers mixed scenarios
- All 455 tests pass

✅ **Backward Compatibility**
- Documents with explicit IDs continue to work
- Auto-generation is seamless
- No breaking changes

✅ **Documentation**
- Schema documentation updated
- AGENTS.md updated
- Code comments are clear and helpful
- Example file added

✅ **Bug Fix Quality**
- Movement detection now correctly handles marker changes
- DiffResult properly populated with all fields
- Title and content changes tracked separately

### Recommendation

**Approve and merge** after fixing the outdated comment on line 439. All other observations are informational and don't require changes.

The implementation is solid, well-tested, and maintains backward compatibility. Great work! 🎉

**Comment by noamoss** [comment]
## 📋 Review Feedback Addressed (Second Review)

I've applied the following change based on the second code review:

### ✅ Changes Made:
- **Fixed outdated comment** on line 439 ([second review](https://github.com/noamoss/yaml_diffs/pull/61#pullrequestreview-3521155976))
  - Updated comment from "Find moved sections (same marker, different path)" 
  - To: "Find moved sections (different path, possibly different marker)"
  - This correctly reflects that the function now handles marker changes

### ✅ Validation:
- All tests passed
- Code formatting verified
- Changes committed locally

### 📦 Next Step:
Pushing changes to remote branch...

---
Ready for final review! 🚀
